app/src/org/commcare/dalvik/activities/PinAuthenticationActivity.java

```diff
@@ -0,0 +1,156 @@
+        currentRecord = CommCareApplication._().getRecordForCurrentUser();
+        if (currentRecord.hasPinSet()) {
+            // If a PIN is already set and the user is trying to change it, we can have them
+            // enter that, and then use it to get the password
+            authMode = LoginMode.PIN;
+        } else {
+            // Otherwise, we're going to need them to enter their password
+            authMode = LoginMode.PASSWORD;
+        }
+        return true;
+    }
+
+    private void setupUI() {
+        if (authMode == LoginMode.PASSWORD) {
+            setPasswordAuthModeUI();
+        } else {
+            setPinAuthModeUI();
+        }
+
+        pinEntry.addTextChangedListener(CreatePinActivity.getPinTextWatcher(enterButton));
+
+        enterButton.setOnClickListener(new View.OnClickListener() {
+
+            @Override
+            public void onClick(View v) {
+                if (authMode == LoginMode.PASSWORD) {
+                    checkEnteredPassword();
+                } else {
+                    checkEnteredPin();
+                }
+            }
+        });
+
+        cancelButton.setOnClickListener(new View.OnClickListener() {
+
+            @Override
+            public void onClick(View v) {
+                setResult(RESULT_CANCELED);
+                finish();
+            }
+        });
+    }
+
+    private void setPinAuthModeUI() {
+        promptText.setText(Localization.get("pin.auth.prompt.pin"));
+        pinEntry.setVisibility(View.VISIBLE);
+        passwordEntry.setVisibility(View.GONE);
+        enterButton.setEnabled(false);
+    }
+
+    private void setPasswordAuthModeUI() {
+        promptText.setText(Localization.get("pin.auth.prompt.password"));
+        passwordEntry.setVisibility(View.VISIBLE);
+        pinEntry.setVisibility(View.GONE);
+        enterButton.setEnabled(true);
+    }
+
+    private void checkEnteredPassword() {
+        String enteredPassword = passwordEntry.getText().toString();
+        if (currentRecord.isPasswordValid(enteredPassword)) {
+            onSuccessfulAuth();
+        } else {
+            onUnsuccessfulAuth();
+        }
+    }
+
+    private void checkEnteredPin() {
+        String enteredPin = pinEntry.getText().toString();
+        if (currentRecord.isPinValid(enteredPin)) {
+            onSuccessfulAuth();
+        } else {
+            onUnsuccessfulAuth();
+        }
+    }
+
+    private void onUnsuccessfulAuth() {
+        if (authMode == LoginMode.PIN) {
+            Toast.makeText(this, Localization.get("pin.auth.failed.pin"), Toast.LENGTH_LONG).show();
+            pinEntry.setText("");
+        } else {
+            Toast.makeText(this, Localization.get("pin.auth.failed.password"), Toast.LENGTH_LONG).show();
+            passwordEntry.setText("");
+        }
+    }
+
+    private void onSuccessfulAuth() {
+        Intent i = new Intent();
+        setResult(RESULT_OK, i);
```

Comment:
I don't think there is a reason to create an `Intent` and set it in the result if you don't put anything in it.